### DISCUSSION TITLE
Dynamic Behavior of Maximum Tokens (max_tokens) in Claude 3 models using LiteLLM

### DISCUSSION THREAD

**Discussion by EnzoAndree** [open-webui/open-webui#1334] [General]
# Bug Report

## Description

**Bug Summary:**
The maximum output token count appears to decrease dynamically, even though the Claude 3 models supports a context window of 200,000 tokens and the max_tokens (Max output) parameter is set to 4096.

**Steps to Reproduce:**
1. Open the Open-WebUI application.
2. Start a new conversation with any Claude 3 model using LiteLLM.
3. Observe the initial `max_tokens` value, which should be 4096 in the log.
4. Continue the conversation by sending additional messages.
5. Notice that the `max_tokens` value decreases with each subsequent message.

**Expected Behavior:**
The `max_tokens` parameter should be kept consistent at 4096 throughout the conversation. It should only be adjusted if the maximum context length is reached (200,000 tokens).

**Actual Behavior:**
The `max_tokens` value decreases as the conversation progresses, even though the context window for the Claude model is 200,000 tokens.

## Environment

- **Operating System:** macOS 14.4
- **Browser (if applicable):** Google Chrome 

## Reproduction Details

**Confirmation:**

- [x] I have read and followed all the instructions provided in the README.md.
- [x] I am on the latest version of both Open WebUI and Ollama.
- [x] I have included the browser console logs.
- [ ] I have included the Docker container logs.

## Logs and Screenshots

```
INFO:     127.0.0.1:56583 - "POST /api/v1/chats/new HTTP/1.1" 200 OK
INFO:     127.0.0.1:56583 - "GET /api/v1/chats/ HTTP/1.1" 200 OK
02:29:38 - LiteLLM:INFO: 

POST Request Sent from LiteLLM:
curl -X POST \
https://api.anthropic.com/v1/messages \
-H 'accept: application/json' -H 'anthropic-version: 2023-06-01' -H 'content-type: application/json' -H 'x-api-key: ********************' \
-d '{'model': 'claude-3-haiku-20240307', 'messages': [{'role': 'user', 'content': 'hello'}], 'stream': True, 'max_tokens': 4085}'


INFO:LiteLLM:

POST Request Sent from LiteLLM:
curl -X POST \
https://api.anthropic.com/v1/messages \
-H 'accept: application/json' -H 'anthropic-version: 2023-06-01' -H 'content-type: application/json' -H 'x-api-key: ********************' \
-d '{'model': 'claude-3-haiku-20240307', 'messages': [{'role': 'user', 'content': 'hello'}], 'stream': True, 'max_tokens': 4085}'


02:29:39 - LiteLLM Router:INFO: litellm.acompletion(model=claude-3-haiku-20240307) 200 OK
INFO:LiteLLM Router:litellm.acompletion(model=claude-3-haiku-20240307) 200 OK
INFO:     127.0.0.1:56583 - "POST /litellm/api/v1/chat/completions HTTP/1.1" 200 OK
INFO:     127.0.0.1:56583 - "POST /api/v1/chats/aa4b137f-2d5a-48f0-810c-0b2567c295bd HTTP/1.1" 200 OK
INFO:     127.0.0.1:56583 - "GET /api/v1/chats/ HTTP/1.1" 200 OK
INFO:     127.0.0.1:56583 - "POST /api/v1/chats/aa4b137f-2d5a-48f0-810c-0b2567c295bd HTTP/1.1" 200 OK
INFO:     127.0.0.1:56583 - "GET /api/v1/chats/ HTTP/1.1" 200 OK
INFO:     127.0.0.1:56583 - "GET /api/v1/chats/ HTTP/1.1" 200 OK
02:30:15 - LiteLLM:INFO: 

POST Request Sent from LiteLLM:
curl -X POST \
https://api.anthropic.com/v1/messages \
-H 'accept: application/json' -H 'anthropic-version: 2023-06-01' -H 'content-type: application/json' -H 'x-api-key: ********************' \
-d '{'model': 'claude-3-haiku-20240307', 'messages': [{'role': 'user', 'content': 'hello'}, {'role': 'assistant', 'content': 'Hello! How can I assist you today?'}, {'role': 'user', 'content': 'this is a test of your max_tokens, please say hello'}], 'stream': True, 'max_tokens': 4063}'


INFO:LiteLLM:

POST Request Sent from LiteLLM:
curl -X POST \
https://api.anthropic.com/v1/messages \
-H 'accept: application/json' -H 'anthropic-version: 2023-06-01' -H 'content-type: application/json' -H 'x-api-key: ********************' \
-d '{'model': 'claude-3-haiku-20240307', 'messages': [{'role': 'user', 'content': 'hello'}, {'role': 'assistant', 'content': 'Hello! How can I assist you today?'}, {'role': 'user', 'content': 'this is a test of your max_tokens, please say hello'}], 'stream': True, 'max_tokens': 4063}'


02:30:16 - LiteLLM Router:INFO: litellm.acompletion(model=claude-3-haiku-20240307) 200 OK
INFO:LiteLLM Router:litellm.acompletion(model=claude-3-haiku-20240307) 200 OK
INFO:     127.0.0.1:56607 - "POST /litellm/api/v1/chat/completions HTTP/1.1" 200 OK
INFO:     127.0.0.1:56607 - "POST /api/v1/chats/aa4b137f-2d5a-48f0-810c-0b2567c295bd HTTP/1.1" 200 OK
INFO:     127.0.0.1:56607 - "GET /api/v1/chats/ HTTP/1.1" 200 OK
INFO:     127.0.0.1:56607 - "GET /api/v1/chats/ HTTP/1.1" 200 OK

```

## Installation Method

Manual installation

**Comment by tjbck** [comment]
This might be a LiteLLM issue, so you might want to try testing LiteLLM in isolation. Keep us updated!

**Reply by EnzoAndree** [reply]
Hi tjbck,

When using LiteLLM, you need to set the max_tokens parameter to 4096 each time you send a message. Otherwise, the output length might get truncated. When I add the model to the Open-WebUI, I set max_tokens to 4096, and that value shouldn't be modified by the application.

I believe that Open-WebUI is trying to manage max_tokens as the maximum context length, but that's not what max_tokens controls. The max_tokens parameter determines the maximum output length for each response, not the entire context window.

The context window for the Claude 3 models is 200,000 tokens, and that should be managed separately from the max_tokens setting. Open-WebUI should not modify the max_tokens value based on the conversation length, as it's intended to control the output length for each individual response.

I hope this clarifies the issue.

**Reply by EnzoAndree** [reply]
I tried the version 1.34.12 of LiteLLM and it seems to be working correctly now. The max_tokens value remains at 4096 throughout the conversation, as expected. It appears the issue was specific to an older version of the library. Thank you for your help in identifying the problem - upgrading to the latest LiteLLM version has resolved the dynamic max_tokens reduction that I was observing before.
